tests/jballerina-unit-test/src/test/resources/test-src/types/uniontypes/union_types_basic.bal

```diff
@@ -325,6 +325,63 @@ function testUnionWithDecimalFiniteTypes() {
     assertEquality(4.3d, <decimal>value + 2.3d);
 }
 
+type UnionTypeWithArray1 int|UnionTypeWithArray1[];
+type intArray int[];
+type UnionTypeWithArray3 UnionTypeWithArray1|string|intArray;
```

Comment:
Let's also add some more tests to check whether it works when a union contains many different types (i.e. xml, map, table, array, tuple, unions of these types, etc) and it's recursive.